File: Network.hs

module Network where

import qualified Data.Set as Set

import Space

data Node = Node
  { nPos    :: Pos
  , nName   :: String
  } deriving (Eq, Ord, Read, Show)

data Edge = Edge
  { eFrom :: Pos
  , eTo   :: Pos
  } deriving (Eq, Ord, Read, Show)

data Symbol = Symbol
  { sColor :: Color
  , sText  :: String
  , sShape :: Shape
  } deriving (Eq, Ord, Read, Show)

data Shape = Circle | Square | Diamond | Hexagon | UpTriangle | DownTriangle
  deriving (Enum, Eq, Ord, Read, Show)

data Color = Color
  { cRed   :: Int
  , cGreen :: Int
  , cBlue  :: Int
  } deriving (Eq, Ord, Read, Show)

data Line = Line
  { lEdges  :: Set.Set Edge
  , lName   :: String
  , lSymbol :: Symbol
  , lColor  :: Color
  } deriving (Eq, Ord, Read, Show)

data Network = Network
  { nNodes     :: Set.Set Node
  , nLines     :: [Line]
  , nLineCount :: Int
  } deriving (Eq, Ord, Read, Show)

File: network.js

const Node = Struct("Node", [['pos', Pos], ['name', String]], function() { });

const Edge = Struct("Edge", [['from', Pos], ['to', Pos]], function() { });

const Color = Struct(
    "Color", [['red', Int], ['green', Int], ['blue', Int]], function() { });

const Shape = Enum(
    "Shape", ['Circle', 'Square', 'Diamond', 'Hexagon', 'UpTriangle', 'DownTriangle']);

const Symbol = Struct(
    "Symbol", [['color', Color], ['text', String], ['shape', Shape]], function() { });

const Line = Struct(
    "Line",
    [['edges', ListOf(Edge)],
     ['name', String],
     ['symbol', Symbol],
     ['color', Color]],
    function() { });

const Network = Struct(
    "Network",
    [['nodes', ListOf(Node)],
     ['lines', ListOf(Line)],
     ['lineCount', Int]],
    function() { });

